feat: Course Unit Screen and Component Navigation Revamp

farhan-arshad-dev opened the pull request:
### Description

[LEARNER-9299](https://2u-internal.atlassian.net/browse/LEARNER-9299)

- Merge `CourseBaseActivity` into `CourseUnitNavigationActivity` cuz `CourseBaseActivity` is not used as a base class for other classes.
- Course Unit Screen and Component Navigation Revamp.

### Notes
- To enable the new Course unit navigation need to set `NEW_DASHBOARD_ENABLED`, `NEW_COURSE_UNIT_NAVIGATION_ENABLED` as `true` in the config.

### Testing
- Legacy Courseunit navigation shouldn't be changed, and work as an app in the production.

### need to add SS for the new design.

omerhabib26 reviewed (changes requested):
1- Can we implement the animation as shown in the video?
https://github.com/openedx/edx-app-android/assets/30689349/f9476bc8-5103-48d3-b0e7-2bd91d1cd8e6
2- Tapping on dropdown popup recreates the popup again n again instead of hiding it.
3- why the Dropdown popup is visible for single components?
4- No design for flying messages? (show a flying message for download enable videos etc over wifi and mobile data)
5- Scrolling behavior is not properly handled for VideoComponents and Discussions
6- NewCourseUnitNavigationActivity misses a lot of old code please verify.


Code Improvements:
It will be great if we can add font weight-age as a comment in [attrs.xml](https://github.com/openedx/edx-app-android/blob/master/OpenEdXMobile/res/values/attrs.xml#L30)

farhan-arshad-dev commented:
> 1- Can we implement the animation as shown in the video?
https://github.com/openedx/edx-app-android/assets/30689349/f9476bc8-5103-48d3-b0e7-2bd91d1cd8e6

Just need to replace the the visibility from `GONE` to `INVISIBLE`.

> 2- Tapping on dropdown popup recreates the popup again n again instead of hiding it.

Using the `PopupWindow` acts like the dialog, so IMO better to recreate the popop (this also helps to refresh the component status in real time.)

> 3- why the Dropdown popup is visible for single components?

Fixed.

>4- No design for flying messages? (show a flying message for download enable videos etc over wifi and mobile data)

Added (missed while rebasing).

>5- Scrolling behavior is not properly handled for VideoComponents and Discussions

Video Component as Discussion screen are using `ListView` that not supported `NestedScrollingChild`. So have to replace with `RecyclerView`. [reference](https://stackoverflow.com/a/33530135)

> 6- NewCourseUnitNavigationActivity misses a lot of old code please verify.

Updated.

>Code Improvements:
?It will be great if we can add font weight-age as a comment in [attrs.xml](https://github.com/openedx/edx-app-android/blob/master/OpenEdXMobile/res/values/attrs.xml#L30)

Comments Added.

omerhabib26 commented on OpenEdXMobile/res/layout/layout_unit_drop_down_item.xml:
```diff
@@ -0,0 +1,46 @@
+    <RelativeLayout
+        android:id="@+id/rl_content"
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/course_unit_nav_spinner_item_height">
```
we can write recyclerView inside a CardView add a corner radius to the parent view and set the background of the item.

omerhabib26 commented on OpenEdXMobile/res/layout/layout_unit_drop_down_item.xml:
```diff
@@ -0,0 +1,64 @@
+            android:layout_weight="1"
+            android:text="@string/locked_content_message_in_drop_down" />
+    </LinearLayout>
+</LinearLayout>
```
Please add a blank line.

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/model/course/CourseComponent.java:
```diff
@@ -579,4 +580,18 @@ public ArrayList<SectionRow> getSectionData() {
         return sections;
     }
+
+    @Nullable
```
IMO, 
```
    /**
     * Method to retrieve the first in-complete child
     * Or return last child if all child are completed
     * Or null if it has 0 child
     */

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/UnitsDropDownAdapter.kt:
```diff
@@ -0,0 +1,64 @@
+
+    override fun onBindViewHolder(holder: UnitDropDownViewHolder, position: Int) {
+        val unit = units[position]
+        holder.binding.tvUnitTitle.text = unit.displayName
```
we can further write it as 
```
        holder.binding.apply {
            tvUnitTitle.text = unit.displayName
            ivUnitStatus.setInVisible(unit.isCompleted.not())
            containerLockedUnit.setVisibility(
                isUpgradeable && unit is CourseComponent &&
                        unit.authorizationDenialReason == AuthorizationDenialReason.FEATURE_BASED_ENROLLMENTS
            )
            rlContent.isSelected = selectedItemPosition == position
            rlContent.setOnClickListener {
                listener?.onItemClick(unit)
            }
        }

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/UnitsDropDownAdapter.kt:
```diff
@@ -0,0 +1,64 @@
+        notifyDataSetChanged()
+    }
+
+    override fun getItemCount(): Int {
```
we can write these method into single line

omerhabib26 commented on OpenEdXMobile/src/test/java/org/edx/mobile/view/CourseBaseActivityTest.java:
```diff
@@ -99,22 +100,22 @@ public void initializeTest() {
-            activity.finishProcess();
-            assertThat(progressWheel).isNotVisible();
-        }
+//        if (progressWheel == null) {
```
any reason to comment out this block?

omerhabib26 commented on OpenEdXMobile/res/layout/activity_course_unit_navigation.xml:
```diff
@@ -0,0 +1,196 @@
+            </RelativeLayout>
+        </androidx.coordinatorlayout.widget.CoordinatorLayout>
+
+        <RelativeLayout
```
IMO, better to replace it with LinearLayout.

omerhabib26 commented on OpenEdXMobile/res/layout/activity_course_unit_navigation.xml:
```diff
@@ -0,0 +1,196 @@
+                                android:layout_marginTop="@dimen/edx_x_quarter_margin">
+
+                                <com.google.android.material.textview.MaterialTextView
+                                    android:id="@+id/course_unit_title"
```
any reason to write this inside framelayout?

omerhabib26 commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseDatesPageFragment.kt:
```diff
@@ -82,15 +84,9 @@ class CourseDatesPageFragment : OfflineSupportBaseFragment() {
-                    false
-                )
-                courseUnitDetailLauncher.launch(courseUnitDetailIntent)
+                showUnitDetailScreen(component)
```
either revert this or move the analytics within the method. 
do the same with else condition.

omerhabib26 reviewed (changes requested):
- Resume Course button isn't working as expected.
Otherwise, there are some minor code improvements.

omerhabib26 commented on OpenEdXMobile/res/values/dimens.xml:
```diff
@@ -154,6 +154,8 @@
     <dimen name="x_small_icon_margin">5dp</dimen>
     <dimen name="small_icon_size">15dp</dimen>
     <dimen name="medium_icon_size">35dp</dimen>
+    <dimen name="segmented_progressbar_height">5dp</dimen>
```
As per design, please update it to 6dp

omerhabib26 reviewed (changes requested):
Only some minor changes.
